packages/flutter/lib/src/widgets/modal_barrier.dart

```diff
@@ -49,20 +50,30 @@ class ModalBarrier extends StatelessWidget {
   ///    [ModalBarrier] built by [ModalRoute] pages.
   final bool dismissible;
 
+  /// Semantics label used for the barrier if it is [dismissable].
+  ///
+  /// See also:
+  ///
+  ///  * [ModalRoute.barrierLabel], which controls this property for the
+  ///    [ModalBarrier] built by [ModalRoute] pages.
+  final String semanticsLabel;
+
   @override
   Widget build(BuildContext context) {
+    final bool semanticsDismissable = dismissible && defaultTargetPlatform != TargetPlatform.android;
     return new BlockSemantics(
       child: new ExcludeSemantics(
         // On Android, the back button is used to dismiss a modal.
-        excluding: !dismissible || defaultTargetPlatform == TargetPlatform.android,
-        child: new Semantics(
-          container: true,
-          child: new GestureDetector(
-            onTapDown: (TapDownDetails details) {
-              if (dismissible)
-                Navigator.pop(context);
-            },
-            behavior: HitTestBehavior.opaque,
+        excluding: !semanticsDismissable,
+        child: new GestureDetector(
+          onTapDown: (TapDownDetails details) {
+            if (dismissible)
+              Navigator.pop(context);
+          },
+          behavior: HitTestBehavior.opaque,
+          child: new Semantics(
+            label: semanticsDismissable ? semanticsLabel : null,
+            textDirection: semanticsDismissable ? Directionality.of(context) : null,
```

Comment:
start the build method with an assert that calls debugCheckHasDirectionality or whatever it's called